text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+from what will be generated compilation time, since (often) an
+expression like `f = (x, y, z)` will wind up desugared into a series
+of piecewise assignments like:
+
+    f.0 = x;
+    f.1 = y;
+    f.2 = z;
+
+However, there are good reasons to include aggregates as first-class
+rvalues. For one thing, if we break down each aggregate into the
+specific assignments that would be used to construct the value, then
+zero-sized types are *never* assigned, since there is no data to
+actually move around at runtime. This means that the compiler couldn't
+distinguish uninitialized variables from initialized ones. That is,
+code like this:
+
+```rust
+let x: (); // note: never initialized
+use(x)
+```
+
+and this:
+
+```rust
+let x: () = ();
+use(x);
+```
+
+would desugar to the same MIR. That is a problem, particularly with
+respect to destructors: imagine that instead of the type `()`, we used
+a type like `struct Foo;` where `Foo` implements `Drop`.
+
+Another advantage is that building aggregates in a two-step way
+assures the proper execution order when unwinding occurs before the
+complete value is constructed. In particular, we want to drop the
+intermediate results in the order that they appear in the source, not
+in the order in which the fields are specified in the struct
+definition.
+
+A final reason to include aggregates is that, at runtime, the
+representation of an aggregate may indeed fit within a single word, in
+which case making a temporary and writing the fields piecemeal may in
+fact not be the correct representation.
+
+In any case, after the move and correctness checking is done, it is
+easy enough to remove these aggregate rvalues and replace them with
+assignments. This could potentially be done during LLVM lowering, or
+as a pre-pass that transforms MIR statements like:
+
+    x = ...x;
+    y = ...y;
+    z = ...z;
+    f = (x, y, z)
+    
+to:
+
+    x = ...x;
+    y = ...y;
+    z = ...z;
+    f.0 = x;
+    f.1 = y;
+    f.2 = z;
+
+combined with another pass that removes temporaries that are only used
+within a single assignment (and nowhere else):
+
+    f.0 = ...x;
+    f.1 = ...y;
+    f.2 = ...z;
+    
+Going further, once type-checking is done, it is plausible to do
+further lowering within the MIR purely for optimization purposes. For
+example, we could introduce intermediate references to cache the
+results of common lvalue computations and so forth. This may well be
+better left to LLVM (or at least to the lowering pass).
+
+### Bounds checking
+
+Because bounds checks are fallible, it's important to encode them in
+the MIR whenever we do indexing. Otherwise the trans code would have
+to figure out on its own how to do unwinding at that point. Because
+the MIR doesn't "desugar" fat pointers, we include a special rvalue
+`LEN` that extracts the length from an array value whose type matches
+`[T]` or `[T;n]` (in the latter case, it yields a constant). Using
```

Comment:
I don't think it complicates the MIR at all. If anything it's simpler as it doesn't require an extra rule for fixed-size arrays. Also, we still need to bounds-check fixed-size arrays, so this would have to be a separate path for them for no obvious reason.